[BUG]게시글 목록 조회 시 이미지 서브쿼리 다중 결과 에러 해결 (.limit(1) 적용)

## 🐞 버그 내용

* **게시글 목록 조회(`findLatestPosts`) API 호출 시 500 Internal Server Error 발생**
* 특정 유저나 게시글 데이터가 로딩될 때 DB 레벨에서 쿼리 실행이 실패함.
* **핵심 에러 로그:** `ERROR: more than one row returned by a subquery used as an expression` (서브쿼리가 2개 이상의 행을 반환함)

```text
2025-12-26T... ERROR 1 --- [http-nio-8080-exec-1] ... : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception ...
Caused by: org.hibernate.exception.DataException: JDBC exception executing SQL [select ... (중략) ... from post p1_0 ...]
    ...
Caused by: org.postgresql.util.PSQLException: ERROR: more than one row returned by a subquery used as an expression
    at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2734)
    ...
    at core.domain.post.repository.impl.PostRepositoryImpl.findLatestPosts(PostRepositoryImpl.java:126)

```

---

## 🧩 예상 원인

* **`PostRepositoryImpl`의 QueryDSL 서브쿼리(`JPAExpressions`) 로직 미흡**
* 게시글 목록 조회 시 **유저 프로필 이미지**나 **게시글 대표 이미지**를 가져오는 서브쿼리가 작성되어 있음.
* 코드상에는 단일 결과(URL 1개)를 기대했으나, 실제 **DB `image` 테이블에 중복 데이터(예: 한 유저의 프로필 사진이 2개 이상)**가 존재하여 서브쿼리가 여러 행을 반환하면서 SQL 에러 발생.

---

## ✅ 재현 경로

1. DB `image` 테이블 조작: 특정 유저(`related_id`)의 `image_type='PROFILE'`(또는 게시글 이미지) 데이터를 강제로 2개 이상 생성.
2. 메인 피드 또는 해당 유저의 글이 포함된 게시글 목록 페이지 접속.
3. `findLatestPosts` 메서드가 실행되는 시점에 **500 Error** 발생 및 서버 로그에 위 에러 출력.
4. **기대 결과:** 데이터가 중복되어 있더라도 최신 이미지 1개만 가져와서(`limit 1`) 정상적으로 리스트가 조회되어야 함.

---

## 📌 이슈 내용

* `PostRepositoryImpl` 게시글 목록 조회 시 이미지 서브쿼리 중복 데이터 예외 처리 (`limit 1` 적용)

---

## ✅ 작업 내용

* [ ] **`PostRepositoryImpl.java` (L126 근처) 수정**
* 이미지 조회 서브쿼리에 `.limit(1)` 추가
* 최신 이미지를 우선하기 위해 `.orderBy(image.createdAt.desc())` (또는 ID 역순) 추가


* [ ] (권장) DB 내 불필요한 중복 이미지 데이터 확인 및 정리

---

## 🔗 관련 이슈/PR

* `PostRepositoryImpl.java`

--- a/src/main/java/core/domain/post/repository/impl/PostRepositoryImpl.java
+++ b/src/main/java/core/domain/post/repository/impl/PostRepositoryImpl.java
@@ -260,6 +260,8 @@ public class PostRepositoryImpl implements PostRepositoryCustom {
                                         userImage.imageType.eq(IMAGE_TYPE_USER)
                                                 .and(userImage.relatedId.eq(user.id))
                                 )
+                                .orderBy(userImage.id.desc())
+                                .limit(1)
                 );
 
         QImage image = QImage.image;
@@ -610,29 +612,23 @@ public class PostRepositoryImpl implements PostRepositoryCustom {
                 .where(
                         u.imageType.eq(IMAGE_TYPE_USER)
                                 .and(u.relatedId.eq(user.id))
-                );
+                )
+                .orderBy(u.id.desc())
+                .limit(1);
     }
 
     private Expression<String> firstPostImageUrlExpr() {
-        QImage pi1 = new QImage("pi1");
-        QImage pi2 = new QImage("pi2");
+        QImage pi = new QImage("pi");
 
         return JPAExpressions
-                .select(pi2.url)
-                .from(pi2)
+                .select(pi.url)
+                .from(pi)
                 .where(
-                        pi2.imageType.eq(IMAGE_TYPE_POST),
-                        pi2.relatedId.eq(post.id),
-                        pi2.id.eq(
-                                JPAExpressions
-                                        .select(pi1.id.min())
-                                        .from(pi1)
-                                        .where(
-                                                pi1.imageType.eq(IMAGE_TYPE_POST),
-                                                pi1.relatedId.eq(post.id)
-                                        )
-                        )
-                );
+                        pi.imageType.eq(IMAGE_TYPE_POST),
+                        pi.relatedId.eq(post.id)
+                )
+                .orderBy(pi.orderIndex.asc(), pi.id.asc())
+                .limit(1); // 첫 번째 이미지 1장만
     }
 
     private Expression<Long> commentCountExpr() {

--- a/src/main/java/core/global/entity/like/entity/Like.java
+++ b/src/main/java/core/global/entity/like/entity/Like.java
@@ -17,7 +17,8 @@ import lombok.NoArgsConstructor;
 @Getter
 @NoArgsConstructor
 public class Like {
-    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     @Column(name = "likes_id")
     private Long id;
 
@@ -25,6 +26,7 @@ public class Like {
     @JoinColumn(name = "user_id", nullable = false)
     private User user;
 
+    @Enumerated(EnumType.STRING)
     @Column(name = "type", nullable = false)
     private LikeType type;
 

--- a/src/main/resources/db/migration/V202512261730__change_image_not_duplicate.sql
+++ b/src/main/resources/db/migration/V202512261730__change_image_not_duplicate.sql
@@ -1,0 +1,3 @@
+ALTER TABLE public.image ADD CONSTRAINT uq_image_type_related_id_order UNIQUE (image_type, related_id, order_index);
+
+ALTER TABLE public.likes ADD CONSTRAINT uq_likes_user_type_related UNIQUE (user_id, type, related_id);

